### PR TITLE
Issue 309 5 and 8

### PR DESCRIPTION
**Pull request by vishwesh3**
PR addressing issue #309 5th and 8th.

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@vishwesh3, it looks like you've already signed the Dropbox CLA. Thanks!

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java [hunk 1]
```diff
@@ -72,7 +73,7 @@ public String getTitle() {
 
     @Override
     public String getSubtitle() {
-        SimpleDateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy");
```
Comment (kunall17) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Why this change here?

File: app/src/main/java/com/zulip/android/util/RemoveViewsOnScroll.java [hunk 2]
```diff
@@ -72,17 +76,29 @@ else if (changeInYDir < 0 && child.getVisibility() == View.GONE && !mIsShowing)
...
         }
     }
 
+    private boolean isTopSnackBar(View child){
+        return (child.getId()!=R.id.appBarLayout && !(child instanceof FloatingActionButton));
+    }
+
     @SuppressLint("NewApi")
     private void hideView(final View view) {
         isViewHidden = true;
```
Comment (kunall17) [inline app/src/main/java/com/zulip/android/util/RemoveViewsOnScroll.java]: code formatting required here!

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java (hunk 1 above)
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Because `./gradlew lintDebug` was giving warning. Now I have removed it. Will be added once https://github.com/zulip/zulip-android/pull/311/ gets merged. :)

File: app/src/main/java/com/zulip/android/util/RemoveViewsOnScroll.java (hunk 2 above)
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/util/RemoveViewsOnScroll.java]: Done

File: app/src/main/java/com/zulip/android/util/RemoveViewsOnScroll.java [hunk 3]
```diff
@@ -43,25 +52,30 @@ public boolean onStartNestedScroll(CoordinatorLayout coordinatorLayout, View chi
...
+                        if (chatBox.getVisibility() == View.VISIBLE) {
+                            return;
+                        }
+                    }
+                    showView(child);
                 }
             }
-            showView(child);
+        } catch (NullPointerException e) {
+            ZLog.logException(e);
```
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/util/RemoveViewsOnScroll.java]: Done :)

File: app/src/main/java/com/zulip/android/activities/ZulipActivity.java [hunk 4]
```diff
@@ -1623,6 +1619,11 @@ public void clearChatBox() {
     }
 
     @Override
+    public void setLayoutBehaviour(LinearLayoutManager linearLayoutManager, RecyclerMessageAdapter adapter) {
+        setActionBarLayoutBehaviour(linearLayoutManager,adapter);
+        setFabLayoutBehaviour(linearLayoutManager,adapter);
```
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/activities/ZulipActivity.java]: Done :)

File: app/src/main/java/com/zulip/android/util/RemoveViewsOnScroll.java (hunk 3 above)
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/util/RemoveViewsOnScroll.java]: @vishwesh3 since these changes were suggested for the first commit, can you please make these changes in 4e426f81b6b5ebea0e12d40c97fcf0ad6e4d6454

Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/util/RemoveViewsOnScroll.java]: Done :)

File: app/src/main/java/com/zulip/android/activities/ZulipActivity.java [hunk 5]
```diff
@@ -1623,6 +1628,29 @@ public void clearChatBox() {
...
+
+        layoutParams = (CoordinatorLayout.LayoutParams) fab.getLayoutParams();
+        layoutParams.setBehavior(new RemoveViewsOnScroll(linearLayoutManager, adapter));
+        fab.setLayoutParams(layoutParams);
+
+        topSnackBar.setMessagesLayoutManager(linearLayoutManager);
+        topSnackBar.setMessagesAdapter(adapter);
+    }
+
+    private int statusBarHeight(){
```
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/activities/ZulipActivity.java]: Its returning height of the status bar. Now I have updated and removed this method.

File: app/src/main/java/com/zulip/android/activities/ZulipActivity.java [hunk 6]
```diff
@@ -2196,38 +2214,35 @@ private void showSnackbarNotification(Message[] messages) {
...
                 }
             });
         }
-        if (appBarLayout.getVisibility() == View.GONE) {
-            snackBarParams.setMargins(0, 0, 0, 0);
-        } else {
-            snackBarParams.setMargins(0, mToolbarHeightInPx, 0, 0);
-        }
-        snackbar.show();
+        topSnackBar.show((appBarLayout.getVisibility() == View.GONE) ? statusBarHeight() : statusBarHeight() + mToolbarHeightInPx,notificationMessage, TopSnackBar.LENGTH_LONG);
```
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/activities/ZulipActivity.java]: Updated. Saved in a `statusBarHeight` variable. :)

File: app/src/main/java/com/zulip/android/util/RemoveViewsOnScroll.java [hunk 7]
```diff
@@ -35,40 +40,56 @@ public RemoveViewsOnScroll(Context context, AttributeSet attrs) {
...
 
     @SuppressLint("NewApi")
     @Override
-    public void onNestedPreScroll(CoordinatorLayout coordinatorLayout, View child, View target, int dx, int dy, int[] consumed) {
-        if (dy > 0 && changeInYDir < 0 || dy < 0 && changeInYDir > 0) {
-            child.animate().cancel();
-            changeInYDir = 0;
-        }
+    public void onNestedPreScroll(CoordinatorLayout coordinatorLayout, View child, View target, int dx, int dy, int[] consumed) throws NullPointerException {
+        if (linearLayoutManager.findLastCompletelyVisibleItemPosition() < adapter.getItemCount() - 3) {
```
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/util/RemoveViewsOnScroll.java]: Done :)

**Comment by vishwesh3** [comment]
Done with the changes. @kunall17 Please have a look.

**Comment by vishwesh3** [comment]
Update: 	Dismiss TopSnackBar when reached at bottom.

**Comment by kunall17** [comment]
Awesome work! 
Just add some margins to the text in the snackbar (spacing from corners) then we can merge this!

device-2017-02-04-122449

**Comment by vishwesh3** [comment]
Added some space in left of the text.
screenshot from 2017-02-04 12-47-37-cropped

**Comment by vishwesh3** [comment]
@kunall17 Done. Please have a final look. :)

**Comment by kunall17** [comment]
@vishwesh3 thanks looking better! 
@timabbott LGTM! :)

**Comment by timabbott** [comment]
Merged, thanks @vishwesh3!  Does this mean we should close #309 now?

**Comment by vishwesh3** [comment]
Yes, #309 can be closed.

**Comment by vishwesh3** [comment]
Closed #309.

**Comment by timabbott** [comment]
Awesome, thanks @vishwesh3!  This polish stuff is really important :).
